### PR TITLE
Ppapi: plugin support

### PR DESCRIPTION
**Pull request by deepak1556**
Related #912 

Currently flash plugin support is implemented though just as a flag for now, start electron with `--ppapi-flash-path`. Will change the behaviour to `ppapi-plugins-dir` option just like `extra-plugins-dir`.

* PPapi 
 - [ ] ~~change the behaviour of loading plugins~~
 * Flash Plugin
    - [x] Fullscreen
    - [ ] Context Menu
* Test
   - [x] Linux
   - [x] OSX
   - [ ] Windows

Depends on https://github.com/atom/libchromiumcontent/pull/106

### CODE REVIEW

**Comment by deepak1556** [comment]
@zcbenz i couldnt find any viable option to set command line flags for pepper plugins apart from providing them in `package.json`, is it good ? Or do we let users just do `app.commandLine.appendSwitch` ?

**Comment by paulcbetts** [comment]
@deepak1556 Set them in the init script, similar to #1519

**Comment by deepak1556** [comment]
Once we have NaCl support was thinking of a config for the `package.json` like

```
pepperPlugins: {
  'plugin-name': {
    'path': '',
    'version': ''
  }
}
```

but that would just make `package.json` more like a `manifest.json`. For now sticking with commandLine switch as its just one plugin.

File: atom/app/atom_content_client.cc [hunk 1]
```diff
@@ -8,9 +8,46 @@
 #include <vector>
 
 #include "atom/common/chrome_version.h"
+#include "atom/common/options_switches.h"
+#include "base/command_line.h"
+#include "base/strings/string_split.h"
+#include "base/strings/string_util.h"
```
Comment (zcbenz) [inline atom/app/atom_content_client.cc]: The `string_` headers do not seem to be actually used?

File: atom/app/atom_content_client.cc [hunk 2]
```diff
@@ -8,9 +8,46 @@
...
+#include "content/public/common/content_constants.h"
+#include "content/public/common/pepper_plugin_info.h"
+#include "ppapi/shared_impl/ppapi_permissions.h"
 
 namespace atom {
 
+int32 kPepperFlashPermissions = ppapi::PERMISSION_ALL_BITS;
```
Comment (zcbenz) [inline atom/app/atom_content_client.cc]: I think we can remove this definition and use `ppapi::PERMISSION_ALL_BITS` directly.

File: atom/app/atom_content_client.cc [hunk 3]
```diff
@@ -8,9 +8,46 @@
...
+
+namespace {
+
+  content::PepperPluginInfo CreatePepperFlashInfo(const base::FilePath& path,
```
Comment (zcbenz) [inline atom/app/atom_content_client.cc]: We don't indent code in namespaces.

File: atom/app/atom_content_client.cc [hunk 4]
```diff
@@ -27,4 +64,19 @@ void AtomContentClient::AddAdditionalSchemes(
   standard_schemes->push_back("chrome-extension");
 }
 
+void AtomContentClient::AddPepperPlugins(
+  std::vector<content::PepperPluginInfo>* plugins) {
```
Comment (zcbenz) [inline atom/app/atom_content_client.cc]: 4 spaces here.

File: atom/app/atom_content_client.cc [hunk 5]
```diff
@@ -27,4 +64,19 @@ void AtomContentClient::AddAdditionalSchemes(
...
+  auto command_line = base::CommandLine::ForCurrentProcess();
+  auto flash_path = command_line->GetSwitchValueNative(
+      switches::kPpapiFlashPath);
+  if (flash_path.empty())
+    return;
+
+  auto flash_version = command_line->GetSwitchValueASCII(
+      switches::kPpapiFlashVersion);
```
Comment (zcbenz) [inline atom/app/atom_content_client.cc]: Should we check if `flash_version` is empty?

File: atom/browser/atom_browser_client.cc [hunk 6]
```diff
@@ -197,6 +200,13 @@ void AtomBrowserClient::AppendExtraCommandLineSwitches(
   dying_render_process_ = nullptr;
 }
 
+void AtomBrowserClient::DidCreatePpapiPlugin(
+    content::BrowserPpapiHost* browser_host) {
+  browser_host->GetPpapiHost()->AddHostFactoryFilter(
+    scoped_ptr<ppapi::host::HostFactory>(
```
Comment (zcbenz) [inline atom/browser/atom_browser_client.cc]: 4 spaces as indent here.

File: atom/app/atom_content_client.cc (hunk 5 above)
Comment (deepak1556) [inline atom/app/atom_content_client.cc]: its optional info for the plugin used only for url request headers to [check](https://code.google.com/p/chromium/codesearch#chromium/src/content/renderer/pepper/url_request_info_util.cc&cl=GROK&l=161&rcl=1431214825). But if users are loading multiple flash plugin at the same time, they should specify anyway. Should that be strict criteria ?

**Comment by zcbenz** [comment]
> @zcbenz i couldnt find any viable option to set command line flags for pepper plugins apart from providing them in package.json, is it good ? Or do we let users just do app.commandLine.appendSwitch ?

The locations of plugins are different on different platforms, so hard-coding them in `package.json` might not work well. Using `app.commandLine.appendSwitch` is fine, and we should mention the switches in `chrome-command-line-switches.md`.

> Currently flash plugin support is implemented though just as a flag for now, start electron with --ppapi-flash-path. Will change the behaviour to ppapi-plugins-dir option just like extra-plugins-dir.

I think it is fine to just support Flash plugin for now.

File: atom/app/atom_content_client.cc (hunk 1 above)
Comment (deepak1556) [inline atom/app/atom_content_client.cc]: yeah it was residual from previous usage to manually parse `flash_version` or provide one if not.

File: atom/app/atom_content_client.cc (hunk 5 above)
Comment (zcbenz) [inline atom/app/atom_content_client.cc]: It sounds fine to make it optional.

**Comment by zcbenz** [comment]
:sparkle:

**Comment by MikeGCV** [comment]
can't make it to load on windows. I don't know if i'm doing somethin wrong

**Comment by deepak1556** [comment]
@MikeGCV can you provide the snippet that you are using to load the plugin ? Also can you open an issue, its better to discuss there :)

**Comment by minggo** [comment]
@deepak1556 Does it support a `pexe` plugin as Chrome did?

**Comment by deepak1556** [comment]
@minggo nope this PR doesnt add the native client runtime.

**Comment by minggo** [comment]
@deepak1556 Got it, thanks. Does it support all pepper api?

**Comment by deepak1556** [comment]
Yup it should support them. As long as the plugin works in corresponding chrome version, it will work fine in Electron.

**Comment by minggo** [comment]
Got it, thanks.

**Pull request merged**